### ISSUE TITLE
User configurable colour and fill for selected and deselected elements

### ISSUE THREAD

**Issue by csaluski** [animint/animint2#65]
Animint's colour and fill currently allow a user to specify a colour mapping in a manner identical to ggplot2, but in interactive visualizations this can prove limiting. Allowing the user to specify a colour mapping for deselected elements can yield much improved visualization clarity by providing emphasis, as well as allowing an additional dimension of data to be displayed. 

In the vein of the [alpha_off](https://github.com/tdhock/animint2/issues/59) feature, I propose implementing this as the new aesthetic parameters `colour_off` and `fill_off`, and allowing the user to specify these parameters for appropriate geometries at either the geometry or aesthetic level. 

This change requires changes at the compiler and runtime levels, as the colour mapping is generated at compile time and output into the `.tsv` files that are loaded by the runtime Javascript. 

Currently aesthetic colour mappings are computed in the compilation process by the [`scales_add_defaults`](https://github.com/tdhock/animint2/blob/7e9505b426ae668594e62b749fd18692340e9dc5/R/scales-.r#L95-L112) function in the layer processing step, which calls [these functions](https://github.com/tdhock/animint2/blob/ce5e887702623ee84af0f7e402f8582d28ebd9be/R/scale-type.R#L1-L32) to look up the colour mapping gradient functions from the parent environments that are defined [here](https://github.com/tdhock/animint2/blob/ce5e887702623ee84af0f7e402f8582d28ebd9be/R/uu_zxx.r). These then call to the wrapper scale functions defined in [`scale-gradient.r`](https://github.com/tdhock/animint2/blob/ce5e887702623ee84af0f7e402f8582d28ebd9be/R/scale-gradient.r) and [`scale-hue.r`](https://github.com/tdhock/animint2/blob/ce5e887702623ee84af0f7e402f8582d28ebd9be/R/scale-hue.r), which finally call into the actual scale generating functions. 

Creating the functions for `colour_off` and `fill_off` will require creating a new version of each function in `uu_zxx.r`, and mapping it to a scale generating function. Currently the functions in `scale-gradient.r` and `scale-hue.r` are individual per the functions in `uu_zxx.r`, however they share almost all of their code and I believe a refactoring to remove this repetition would be worthwhile, and would prevent the addition of `colour_off` and `fill_off` further increasing the size and repetition of these files.

I have not yet investigated how geometry level colors are assigned. 

Once the compiler has been modified to support these properties implementing them in the Javascript should be fairly simple, and will be very similar to how the `alpha` and `alpha_off` parameters are implemented.

**Comment by Faye-yufan** [comment]
In compiler, I believe is [here](https://github.com/tdhock/animint2/blob/master/R/z_animint.R#L808-L809) allows `color` and `fill` is the same hex color code in the output chunk tsv file. Implementing `color_off` and `fill_off` shall be the same.

In JS, add a new function `colour` like [`opacity`](https://github.com/tdhock/animint2/blob/master/inst/htmljs/animint.js#L1570-L1586) for updating the non-selected elements would work.

**Comment by Faye-yufan** [comment]
Here are a couple of notes on discussing with @tdhock today: 
* sometimes it would not be very clear when the color is already used to distinguish other data property in a legend. So maybe we should consider how to avoid selection state color messing with the legend scaled color. Some thoughts following [this comment](https://github.com/tdhock/animint2/pull/67#discussion_r917021135):  
    * Let `color_off` and `fill_off` only work on single static color, like "grey", or "black". 
    * Sometimes, if the trend of a large set of data is much the same, like the "Surface temperature" example in [this rendered Rmd](https://rcdata.nau.edu/genomic-ml/public_html/animint-paper-figures/), it would be more clear if just keep the selected line and let others disappear. So I'm thinking about adding a new button or checkbox inside `selection menus` in html to allow user clicks and change the color of non-selected variables to the background color.
* If the aesthetic appears in the `aes`, it shall be used for wrapping a data variable and cannot be used for the selection mapping.

**Comment by Faye-yufan** [comment]
@csaluski A small [demo](http://bl.ocks.org/Faye-yufan/raw/a03556b5c4ae261cc0a2112d1694bf73/) on this, but I'm not sure if it makes any sense.

**Comment by csaluski** [comment]
I think the demo is good for demoing the deactivated fill, it's a very useful way to draw some extra emphasis to currently selected items. 

If we're not going to have a deselected color the addition of the extra `colour_off` scales aren't needed at all, and I think the demo has all the features we'll need for it. I agree that it could be confusing to have 2 different color scales in a single figure, but maybe if we did a check to see if the color ramps didn't have any overlap that confusion could be avoided.

**Comment by tdhock** [comment]
hi @Faye-yufan thanks for sharing that example/demo. as I understand it you have grey fill to indicate points which are not selected, and then you have fill of different shades of blue for selected points. I guess that is understandable, but I think it could be highly confusing to use the fill in both ways. If you want to use a fill legend in a plot to represent data values, I think you should use something else (size, color, alpha) to represent selection state, to avoid confusion. Does that make sense?

**Comment by tdhock** [comment]
#72 implements color_off so partially fixes this issue. plan is to implement fill_off in a future pr.

**Comment by tdhock** [comment]
also would be nice to 
- warn if *_off is used on a geom which does not specify clickSelects. something like: geom_point has color_off="transparent" which is not used because this geom has no clickSelects; please specify clickSelects or remove color_off for this geom,
- update animint2 manual to mention these new features. alpha_off can be discussed in this section, https://rcdata.nau.edu/genomic-ml/animint2-manual/Ch04-clickSelects.html#clickSelects-keyword right after the alpha-0.5 default is first mentioned. color_off probably should be mentioned in Ch12 when geom_tile is used to select svm hyper params, https://rcdata.nau.edu/genomic-ml/animint2-manual/Ch12-SVM.html

**Comment by Faye-yufan** [comment]
hi @tdhock , as I looking into the SVM example with _geom_tile_, I found the `aes(colour)`, if not defined, is default set equal to `aes(fill)` in the compiler, see [here](https://github.com/tdhock/animint2/blob/22e4d08ebdcfba9e3df73ec0fa9469094499192d/R/geom-tile.r#L106-108), so as _geom_raster_. I'm wondering why we need `g.data[["colour"]]` here? If I remove this line, it still works.

As the comment above said 

> Color set to match ggplot2 default of tile with no outside border.

maybe we can go without `colour`? Since if we keep it, it might lead to writing more chunky code to change the colour/stroke of _geom_tile/geom_raster_, so as generalize the code on the JS side.

**Comment by tdhock** [comment]
I'm not sure I understand your question. If you remove what line, what still works? If all of the tests still work after that, then that means we should probably add a new test case to clarify this ambiguity.
Can you please clarify "more chunky code" ?
For that SVM example the current/old code is
```r
    geom_tile(aes(
      log10(cost.param), degree.num, fill=errors),
      clickSelects="degree.num",
      data=testErrors),
```
Although the color is not specified, the default implies color="black" and color_off="transparent".
We could demo how to use the new feature by specifying, for example, 
```r
    geom_tile(aes(
      log10(cost.param), degree.num, fill=errors),
      clickSelects="degree.num",
      color_off="grey",
      color="grey50",
      data=testErrors),
```

**Comment by Faye-yufan** [comment]
"it works" means geom_tile still has the `fill` but no `colour`, except user specifies the colour in aes.

For my understanding, `g.data[["colour"]] <- g.data[["fill"]]` would give each tile the same colour value as fill in the final .tsv file, and for now, the JS would take the colour column in .tsv file first and ignore the colour/colour_off parameters, which leads to not able to use _colour_off_ to change the colour/stroke for geom_tile.

"more chunky" means that if we keep both _fill_ and _colour_ columns in .tsv file, we should add more code for this exception to specify the colour/stroke of geom_tile.

**Comment by tdhock** [comment]
does this work?
```r
    geom_tile(aes(
      log10(cost.param), degree.num, fill=errors),
      color="grey50",
      data=testErrors),
```

**Comment by Faye-yufan** [comment]
Seems the `color="grey50"` does not work
![image](https://user-images.githubusercontent.com/62975717/186957751-4a9791ae-5eb7-418d-8257-5838b9f701c7.png)

**Comment by Faye-yufan** [comment]
> also would be nice to
> 
> * warn if *_off is used on a geom which does not specify clickSelects. something like: geom_point has color_off="transparent" which is not used because this geom has no clickSelects; please specify clickSelects or remove color_off for this geom,
> * update animint2 manual to mention these new features. alpha_off can be discussed in this section, https://rcdata.nau.edu/genomic-ml/animint2-manual/Ch04-clickSelects.html#clickSelects-keyword right after the alpha-0.5 default is first mentioned. color_off probably should be mentioned in Ch12 when geom_tile is used to select svm hyper params, https://rcdata.nau.edu/genomic-ml/animint2-manual/Ch12-SVM.html

- warning has been added in the #72  commit
- animint manual update for `alpha_off` can be found in [pr#2](https://github.com/tdhock/animint-book/pull/2), and `colour_off`  in [pr#6](https://github.com/tdhock/animint-book/pull/6)

**Comment by tdhock** [comment]
This PR is still open because fill_off is not yet implemented.

**Comment by Faye-yufan** [comment]
fill_off has been implemented in PR #104 .
